src/common/src/common-string-private.h

```diff
@@ -20,75 +20,537 @@
+ */
 static BSON_INLINE void
-mcommon_string_append_c (mcommon_string_t *string, char str)
+mcommon_string_append_init (mcommon_string_append_t *new_append, mcommon_string_t *string)
 {
-   BEGIN_IGNORE_DEPRECATIONS
-   bson_string_append_c (string, str);
-   END_IGNORE_DEPRECATIONS
+   mcommon_string_append_init_with_limit (new_append, string, UINT32_MAX - 1u);
 }
+
+/**
+ * @brief Begin appending to a new empty mcommon_string_t with default initial capacity and maximum length
+ * @param new_append Pointer to an uninitialized mcommon_string_append_t
+ *
+ * Allocates a new mcommon_string_t, which will need to be deallocated by the caller.
+ * The mcommon_string_append_t itself does not need to be deallocated.
+ *
+ * The maximum string length will be set to the largest representable by the data type, UINT32_MAX - 1.
+ * The new string will be allocated with a small default capacity.
+ *
+ * This method is intended to be the most convenient way to start growing a string. If a reasonable guess
+ * can be made about the final size of the string, it's better to call mcommon_string_new_with_capacity()
+ * and mcommon_string_append_init() directly.
+ */
 static BSON_INLINE void
-mcommon_string_append_unichar (mcommon_string_t *string, bson_unichar_t unichar)
+mcommon_string_append_new (mcommon_string_append_t *new_append)
 {
-   BEGIN_IGNORE_DEPRECATIONS
-   bson_string_append_unichar (string, unichar);
-   END_IGNORE_DEPRECATIONS
+   mcommon_string_append_init (new_append, mcommon_string_new_with_capacity ("", 0, 32));
 }
 
+/**
+ * @brief Begin appending to a new empty mcommon_string_t with a default capacity and a matching max append length.
+ * @param new_append Pointer to an uninitialized mcommon_string_append_t
+ *
+ * Allocates a new mcommon_string_t, which will need to be deallocated by the caller.
+ * The mcommon_string_append_t itself does not need to be deallocated.
+ */
 static BSON_INLINE void
-mcommon_string_append_printf (mcommon_string_t *string, const char *format, ...) BSON_GNUC_PRINTF (2, 3);
+mcommon_string_append_new_fixed_capacity (mcommon_string_append_t *new_append, uint32_t capacity)
+{
+   mcommon_string_append_init_with_limit (new_append, mcommon_string_new_with_capacity ("", 0, capacity), capacity);
+}
+
+/**
+ * @brief Check the status of an append operation
+ * @param append Append operation, initialized with mcommon_string_append_init
+ * @returns true if the append operation has no permanent error status. false if the max length has been exceeded.
+ */
+static BSON_INLINE bool
+mcommon_string_append_status (const mcommon_string_append_t *append)
+{
+   return !append->_max_len_exceeded;
+}
 
+/**
+ * @brief Get a reference to the string being appended to
+ * @param append Append operation, initialized with mcommon_string_append_init
+ * @returns Pointer to the mcommon_string_t being written to
+ */
+static BSON_INLINE mcommon_string_t *
+mcommon_string_append_destination (const mcommon_string_append_t *append)
```

Comment:
I almost added this myself, then decided against it because it seemed wise to nudge code toward using mcommon_string_t more pervasively. I agree it would be a useful shortcut for the existing call sites.